hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/s3guard/ITestS3GuardFsck.java

```diff
@@ -0,0 +1,707 @@
+      awaitFileStatus(rawFS, file);
+
+      final S3GuardFsck s3GuardFsck =
+          new S3GuardFsck(rawFS, metadataStore);
+
+      final List<S3GuardFsck.ComparePair> comparePairs =
+          s3GuardFsck.compareS3ToMs(cwd);
+
+      assertEquals("Number of pairs should be two.", 2,
+          comparePairs.size());
+      final S3GuardFsck.ComparePair pair = comparePairs.get(0);
+      assertTrue("The pair must contain a violation.", pair.containsViolation());
+      assertEquals("The pair must contain only one violation", 1,
+          pair.getViolations().size());
+
+      final S3GuardFsck.Violation violation =
+          pair.getViolations().iterator().next();
+      assertEquals("The violation should be that there is no violation entry.",
+          violation, S3GuardFsck.Violation.NO_METADATA_ENTRY);
+    } finally {
+      // delete the working directory with all of its contents
+      rawFS.delete(cwd, true);
+      metadataStore.forgetMetadata(file);
+      metadataStore.forgetMetadata(cwd);
+    }
+  }
+
+  @Test
+  public void testIDetectNoParentEntry() throws Exception {
+    final Path cwd = path("/" + getMethodName() + "-" + UUID.randomUUID());
+    final Path file = new Path(cwd, "file");
+    try {
+      // create a file with guarded fs
+      touch(guardedFs, file);
+      awaitFileStatus(guardedFs, file);
+
+      // delete the parent from the MS
+      metadataStore.forgetMetadata(cwd);
+
+      final S3GuardFsck s3GuardFsck =
+          new S3GuardFsck(rawFS, metadataStore);
+
+      final List<S3GuardFsck.ComparePair> comparePairs =
+          s3GuardFsck.compareS3ToMs(cwd);
+
+      assertEquals("Number of pairs should be two. The cwd (parent) and the "
+              + "child.", 2, comparePairs.size());
+
+      // check the parent that it does not exist
+      final S3GuardFsck.ComparePair cwdPair = comparePairs.stream()
+          .filter(p -> p.getPath().equals(cwd))
+          .findFirst().get();
+      assertNotNull("The pair should not be null.", cwdPair);
+      assertTrue("The cwdPair must contain a violation.", cwdPair.containsViolation());
+      Assertions.assertThat(cwdPair.getViolations())
+          .describedAs("Violations in the cwdPair")
+          .contains(S3GuardFsck.Violation.NO_METADATA_ENTRY);
+
+      // check the child that there's no parent entry.
+      final S3GuardFsck.ComparePair childPair = comparePairs.stream()
+          .filter(p -> p.getPath().equals(file))
+          .findFirst().get();
+      assertNotNull("The pair should not be null.", childPair);
+      assertTrue("The childPair must contain a violation.", childPair.containsViolation());
+      Assertions.assertThat(childPair.getViolations())
+          .describedAs("Violations in the childPair")
+          .contains(S3GuardFsck.Violation.NO_PARENT_ENTRY);
+    } finally {
+      // delete the working directory with all of its contents
+      rawFS.delete(cwd, true);
+      metadataStore.forgetMetadata(file);
```

Comment:
this cleanup happens in every test. what about adding a list of files to forget in teardown, which can go through each one, forget them, and swallow exceptions raised? Or have a `forget(Path...paths)` method to take a varargs list